R/parseResults.R

```diff
@@ -307,52 +307,104 @@ compute_partial_correlation <- function(summary, observations, state_order) {
+  v_structs <- probas$NI3 < 0 & probas$Error == 0
+  if (!any(v_structs)) {
     return(is_causal_results)
   }
-  probas_V_structs = probas[v_structs,]
-
-  vstruct_matrix <- matrix_from_3_columns(
-    probas_V_structs, "source1",
-    "source2", "target"
-  )
-  error_matrix <- matrix_from_3_columns(
-    probas_V_structs, "source1",
-    "source2", "Error"
-  )
+  probas_v <- probas[v_structs, ]
+  probas_non_v <- probas[probas$NI3 > 0 & probas$Error == 0, ]
 
   for (i in 1:nrow(summary)) {
     row <- summary[i, ]
-    if (row$type %in% c("TN", "FN", "N") || # Negative edge
-      row$infOrt == 1) { # Non oriented edge
-      # If any of these conditions is true, edge cannot be causal
+    # Negative or Non oriented edge cannot be causal
+    if (row$type %in% c("TN", "FN", "N") || row$infOrt == 1) {
       next
     }
+    # Bi-directed edge cannot be causal
     if (row$infOrt == 6) {
-      is_causal_results[i] <- "Y"
       next
     }
-
+    # source --> target (source (tail)-(head) target)
     if (row$infOrt == 2) {
-      source <- row$x
-      target <- row$y
+      source_node <- row$x
+      target_node <- row$y
     } else {
-      source <- row$y
-      target <- row$x
+      source_node <- row$y
+      target_node <- row$x
     }
-
-    if (!any(probas_V_structs[, "target"] == source)) {
-      # Propagated orientation
+    main_v_structs <- probas_v$target == target_node &
+        (probas_v$source1 == source_node | probas_v$source2 == source_node)
+    # Edge is not part of any V-structure (propagated orientation), condition 1.
+    # not satisfied.
+    if (!any(main_v_structs)) {
       next
     }
-
-    # If there is another V-structure pointing to the source node, which
-    # is not an error.
-    if (any(error_matrix[vstruct_matrix == source] == 0)) {
-      is_causal_results[i] <- "Y"
+    second_v_structs <- probas_v$target == source_node
+    # source is not the mid node of any V-structure, condition 2. not satisfied.
+    if (!any(second_v_structs)) {
+      next
+    }
+    probas_second_v_list <- probas_v[second_v_structs, ]
+    for (j in 1:nrow(probas_second_v_list)) {
+      proba_row <- probas_second_v_list[j, ]
+      s1 <- proba_row$source1
+      s2 <- proba_row$source2
+      # s1 --> source_node --> target_node
+      s1_non_v <- probas_non_v$target == source_node &
+          ((probas_non_v$source1 == s1 & probas_non_v$source2 == target_node) |
+          (probas_non_v$source2 == s1 & probas_non_v$source1 == target_node))
+      # s2 --> source_node <-- target_node
+      s2_v <- probas_v$target == source_node &
+          ((probas_v$source1 == s2 & probas_v$source2 == target_node) |
+          (probas_v$source2 == s2 & probas_v$source1 == target_node))
+      # Condition 3.
+      if (any(s1_non_v)) {
+        # Condition 4.
+        # There is at most one true element in s1_non_v and s2_v, still we use
+        # [1, ] to explicitly ensure the correct dimention
```

Comment:
typo

Reply:
I blame auto-completion :P Though the first one was on me too